File: freesia-spring-boot/src/main/java/edu/coldrain/freesia/service/UnitService.java
package edu.coldrain.freesia.service;

import edu.coldrain.freesia.dto.UnitDTO;
import edu.coldrain.freesia.entity.Unit;
import edu.coldrain.freesia.entity.Vocabulary;
import edu.coldrain.freesia.entity.Word;
import edu.coldrain.freesia.exception.UnitNotFoundException;
import edu.coldrain.freesia.exception.VocabularyNotFoundException;
import edu.coldrain.freesia.repository.UnitRepository;
import edu.coldrain.freesia.repository.VocabularyRepository;
import edu.coldrain.freesia.repository.WordRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UnitService {

    private final UnitRepository unitRepository;

    private final VocabularyRepository vocabularyRepository;

    private final WordRepository wordRepository;

    public Page<UnitDTO.Response> searchUnitResponsePage(Long vocabularyId, Pageable pageable) {
        return unitRepository.searchUnitResponsePage(vocabularyId, pageable);
    }

    public UnitDTO.Response searchOneUnitResponse(Long unitId) {
        return unitRepository.searchOneUnitResponse(unitId);
    }

    @Transactional
    public void removeUnit(Long unitId) {
        // delete all children
        final List<Word> words = wordRepository.findAllByParentId(unitId);
        wordRepository.deleteAllInBatch(words);

        // delete parent
        unitRepository.deleteById(unitId);
    }

    @Transactional
    public void modifyUnit(Long unitId, UnitDTO.ModifyRequest request) {
        final Unit unit = unitRepository.findById(unitId)
                .orElseThrow(() -> new UnitNotFoundException("Unit not found exception."));

        unit.changeSubject(request.getSubject());
    }

    public Long registerUnit(Long vocabularyId, UnitDTO.RegistrationRequest request) {
        final Vocabulary vocabulary = vocabularyRepository.findById(vocabularyId)
                .orElseThrow(() -> new VocabularyNotFoundException("Vocabulary not found exception."));

        final Unit unit = Unit.builder()
                .vocabulary(vocabulary)
                .subject(request.getSubject())
                .build();

        final Unit savedUnit = unitRepository.save(unit);
        return savedUnit.getId();
    }

    public Long findRownumById(Long unitId) {
        return unitRepository.findRownumById(unitId);
    }
}

File: freesia-vuetify/src/service/unitService.js
import { $axios } from "@/common/axios";

export const unitService = {
  /** 단위 목록 페이지 조회 API */
  getPageable: async (vocabularyId = 0, pageable = { page: 0, size: 3 }) => {
    const { page, size } = pageable;
    const response = await $axios.get(
      `/vocabulary/${vocabularyId}/units?page=${page}&size=${size}`
    );
    return response.data;
  },

  /** 단위 등록 API */
  register: async (vocabularyId, formData) => {
    const { subject } = formData;
    const response = await $axios.post(`/vocabulary/${vocabularyId}/units`, {
      subject,
    });
    return response.data;
  },

  /** 단위 삭제 API */
  removeById: async (unitId) => {
    await $axios.delete(`/units/${unitId}`);
  },

  /** 단위 수정 API */
  modifyById: async (unitId, formData) => {
    const { subject } = formData;
    await $axios.patch(`/units/${unitId}`, {
      subject,
    });
  },
};
